lib/pbench/test/unit/agent/task/test_tools_kill.py

```diff
@@ -0,0 +1,440 @@
+            and pairs[0][0].name == "tm"
+            and pairs[0][1] == "abcdef"
+        )
+        assert (
+            pairs[1][0].parent.name == "run2"
+            and pairs[1][0].name == "tm"
+            and pairs[1][1] == "ghijkl"
+        )
+
+    @staticmethod
+    def test_gen_host_names(monkeypatch):
+        class MockToolGroup:
+            """A simplified tool group object which only provides the
+            `hostnames` dictionary.
+            """
+
+            def __init__(self, hosts: List[str]):
+                self.hostnames = {}
+                for host in hosts:
+                    self.hostnames[host] = None
+
+        tool_group_host_names = [
+            ["localhost.example.com"],
+            ["localhost.example.com", "remote1.example.com"],
+            ["remote2.example.com", "remote3.example.com"],
+        ]
+        expected_hosts = set(
+            ("remote1.example.com", "remote2.example.com", "remote3.example.com")
+        )
+
+        def mock_gen_tool_groups(pbench_run: str) -> Iterable[MockToolGroup]:
+            """Override kill.gen_tool_groups definition to return a simple
+            tool group object.
+            """
+            for hosts in tool_group_host_names:
+                yield MockToolGroup(hosts)
+
+        class MockLocalRemoteHost:
+            """A simple mock for the LocalRemoteHost behaviors."""
+
+            def is_local(self, host_name: str):
+                return host_name.startswith("localhost")
+
+        # To verify the case where there are no tool groups created, we make a
+        # special Path object which will expose no tool groups.
+        hosts = list(kill.gen_host_names(pathlib.Path("no-tool-group-dirs")))
+        assert len(hosts) == 0
+
+        monkeypatch.setattr(
+            "pbench.cli.agent.commands.tools.kill.gen_tool_groups", mock_gen_tool_groups
+        )
+        monkeypatch.setattr(
+            "pbench.cli.agent.commands.tools.kill.LocalRemoteHost", MockLocalRemoteHost
+        )
+
+        # Using the special Path object which does have tool groups, we
+        # verify that the expected generated list of remote hosts are all
+        # the expected hosts listed in the MockToolGroup class above.
+        hosts = set(kill.gen_host_names(pathlib.Path("have-tool-group-dirs")))
+        assert hosts == expected_hosts
+
+    @staticmethod
+    def test_with_uuids(monkeypatch):
+        """Verify UUID arguments work correctly.
+
+        The scenario we construct artificially starts with an invocation using
+        3 UUIDs, one won't be found.  The mocked-out `psutil.process_iter()`
+        will generate 3 PIDs, one of which will have child processes, which
+        will all operate normally with no exceptions.  Of the other two, one
+        will raise an expected exception, and the other will not be found.
+        """
+        action_list = []
+        uuid_list = ["uuid1abc", "uuid2def", "uuid3ghi"]
```

Comment:
Consider using a(n immutable) tuple here instead of a list.  (It shouldn't affect the code which references `uuid_list`, but it will prevent `uuid_list` from being inadvertently altered.)